src/micro.c

```diff
@@ -0,0 +1,680 @@
+    IFOK(s, natsConn_setDisconnectedCallback(m->nc, m->prev_on_connection_disconnected, m->prev_on_connection_disconnected_closure));
+    IFOK(s, natsConn_setErrorCallback(m->nc, m->prev_on_error, m->prev_on_error_closure));
+
+    return microError_Wrapf(micro_ErrorFromStatus(s), "failed to unwrap connection event callbacks");
+}
+
+microError *
+microService_AddGroup(microGroup **new_group, microService *m, const char *prefix)
+{
+    if ((m == NULL) || (new_group == NULL) || (prefix == NULL))
+        return micro_ErrorInvalidArg;
+
+    *new_group = NATS_CALLOC(1, sizeof(microGroup) +
+                                    strlen(prefix) + 1); // "prefix.""
+    if (new_group == NULL)
+    {
+        return micro_ErrorOutOfMemory;
+    }
+
+    memcpy((*new_group)->prefix, prefix, strlen(prefix) + 1);
+    (*new_group)->m = m;
+    (*new_group)->next = m->groups;
+    m->groups = *new_group;
+
+    return NULL;
+}
+
+microError *
+microGroup_AddGroup(microGroup **new_group, microGroup *parent, const char *prefix)
+{
+    char *p;
+    int len;
+
+    if ((parent == NULL) || (new_group == NULL) || (prefix == NULL))
+        return micro_ErrorInvalidArg;
+
+    *new_group = NATS_CALLOC(1, sizeof(microGroup) +
+                                    strlen(parent->prefix) + 1 + // "parent_prefix."
+                                    strlen(prefix) + 1);         // "prefix\0"
+    if (new_group == NULL)
+    {
+        return micro_ErrorOutOfMemory;
+    }
+
+    p = (*new_group)->prefix;
+    len = strlen(parent->prefix);
+    memcpy(p, parent->prefix, len);
+    p[len] = '.';
+    p += len + 1;
+    memcpy(p, prefix, strlen(prefix) + 1);
+    (*new_group)->m = parent->m;
+    (*new_group)->next = parent->m->groups;
+    parent->m->groups = *new_group;
+
+    return NULL;
+}
+
+natsConnection *
+microService_GetConnection(microService *m)
+{
+    if (m == NULL)
+        return NULL;
+    return m->nc;
+}
+
+microError *
+microService_GetInfo(microServiceInfo **new_info, microService *m)
+{
+    microServiceInfo *info = NULL;
+    microEndpoint *ep = NULL;
+    int len = 0;
+
+    if ((new_info == NULL) || (m == NULL) || (m->service_mu == NULL))
+        return micro_ErrorInvalidArg;
+
+    info = NATS_CALLOC(1, sizeof(microServiceInfo));
+    if (info == NULL)
+        return micro_ErrorOutOfMemory;
+
+    info->Name = m->cfg->Name;
+    info->Version = m->cfg->Version;
+    info->Description = m->cfg->Description;
+    info->Id = m->id;
+    info->Type = MICRO_INFO_RESPONSE_TYPE;
+
+    // Overallocate subjects, will filter out internal ones.
+    info->Subjects = NATS_CALLOC(m->num_eps, sizeof(char *));
+    if (info->Subjects == NULL)
+    {
+        NATS_FREE(info);
+        return micro_ErrorOutOfMemory;
+    }
+
+    micro_lock_service(m);
+    for (ep = m->first_ep; ep != NULL; ep = ep->next)
+    {
+        if ((!ep->is_monitoring_endpoint) && (ep->subject != NULL))
+        {
+            info->Subjects[len] = ep->subject;
```

Comment:
We are not strdup'ing those, which is fine as long as it is documented that strings belong to the micro service and should not be altered/freed by the user.

Reply:
This is a mistake, it's a public API, these should be duped. Fixing.

Reply:
Looking at it, they are declared as `const char **`, which (perhaps again a #backfromgoland) I thought was a `(const char *) *`, which would imply the correct semantics. Am I wrong? (still, since a public API, better to dupe/free them).